File: MatriX/samples/vb/MiniClient/FrmGroupChat.vb
Imports System.Drawing
Imports System.Windows.Forms
Imports Matrix
Imports Matrix.Xmpp
Imports Matrix.Xmpp.Client

Public Partial Class FrmGroupChat
    Inherits Form
    Private mm As MucManager
		
    Public Sub New(xmppClient As XmppClient, roomJid As Jid, nickname As String)
        InitializeComponent()

        _roomJid = roomJid
        _xmppClient = xmppClient
        _nickname = nickname

        mm = New MucManager(xmppClient)

        ' Setup new Message Callback using the MessageFilter
        _xmppClient.MessageFilter.Add(roomJid, New BareJidComparer(), AddressOf MessageCallback)

        ' Setup new Presence Callback using the PresenceFilter

        _xmppClient.PresenceFilter.Add(roomJid, New BareJidComparer(),AddressOf PresenceCallback)
    End Sub
		
    Private ReadOnly _roomJid As Jid
    Private ReadOnly _xmppClient As XmppClient
    Private ReadOnly _nickname As String

    Private Sub FrmGroupChat_Load(sender As Object, e As System.EventArgs) Handles MyBase.Load
        If _roomJid IsNot Nothing Then
            mm.EnterRoom(_roomJid, _nickname)
        End If
    End Sub

    Private Sub FrmGroupChat_FormClosed(sender As Object, e As FormClosedEventArgs) Handles MyBase.FormClosed
        If _roomJid IsNot Nothing Then
            mm.ExitRoom(_roomJid, _nickname)

            ' Remove the Message Callback in the MessageFilter
            _xmppClient.MessageFilter.Remove(_roomJid)

            ' Remove the Presence Callback in the MessageFilter
            _xmppClient.PresenceFilter.Remove(_roomJid)
        End If

    End Sub

    Private Sub MessageCallback(sender As Object, e As MessageEventArgs)
        If e.Message.Type = MessageType.groupchat Then
            IncomingMessage(e.Message)
        End If
    End Sub


    Private Sub PresenceCallback(sender As Object, e As PresenceEventArgs)

        Dim mucX = e.Presence.MucUser

        ' check for status code 201, this means the room is not ready to use yet
        ' we request an instant room and accept the and accept the default configuration by the server
        If mucX.HasStatus(201) Then
            ' 201 =  room is awaiting configuration.
            mm.RequestInstantRoom(_roomJid)
        End If


        Dim lvi = FindListViewItem(e.Presence.From)
        If lvi IsNot Nothing Then
            If e.Presence.Type = PresenceType.unavailable Then
                lvi.Remove()
            Else
                Dim imageIdx As Integer = Util.GetRosterImageIndex(e.Presence)
                lvi.ImageIndex = imageIdx
                lvi.SubItems(1).Text = (If(e.Presence.Status, ""))

                Dim u = e.Presence.MucUser
                If u IsNot Nothing Then
                    lvi.SubItems(2).Text = u.Item.Affiliation.ToString()
                    lvi.SubItems(3).Text = u.Item.Role.ToString()
                End If
            End If
        Else
            Dim imageIdx As Integer = Util.GetRosterImageIndex(e.Presence)

            Dim lv = New ListViewItem(e.Presence.From.Resource) With { _
                    .Tag = e.Presence.From.ToString() _
                    }

            lv.SubItems.Add(If(e.Presence.Status, ""))

            Dim u = e.Presence.MucUser
            If u IsNot Nothing Then
                lv.SubItems.Add(u.Item.Affiliation.ToString())
                lv.SubItems.Add(u.Item.Role.ToString())
            End If
            lv.ImageIndex = imageIdx
            lvwRoster.Items.Add(lv)
        End If
    End Sub

    Private Function FindListViewItem(jid As Jid) As ListViewItem
        For Each lvi As ListViewItem In lvwRoster.Items
            If jid.ToString().ToLower() = lvi.Tag.ToString().ToLower() Then
                Return lvi
            End If
        Next
        Return Nothing
    End Function

    Private Sub IncomingMessage(msg As Matrix.Xmpp.Client.Message)
        If msg.Type = MessageType.[error] Then
            'Handle errors here
            ' we dont handle them in this example
            Return
        End If

        If msg.Subject IsNot Nothing Then
            txtSubject.Text = msg.Subject

            rtfChat.SelectionColor = Color.DarkGreen
            ' The Nickname of the sender is in GroupChat in the Resource of the Jid
            rtfChat.AppendText(msg.From.Resource + " changed subject: ")
            rtfChat.SelectionColor = Color.Black
            rtfChat.AppendText(msg.Subject)
            rtfChat.AppendText(vbCr & vbLf)
        Else
            If msg.Body Is Nothing Then
                Return
            End If

            rtfChat.SelectionColor = Color.Red
            ' The Nickname of the sender is in GroupChat in the Resource of the Jid
            rtfChat.AppendText(msg.From.Resource + " said: ")
            rtfChat.SelectionColor = Color.Black
            rtfChat.AppendText(msg.Body)
            rtfChat.AppendText(vbCr & vbLf)
        End If
    End Sub

    Private Sub cmdSend_Click(sender As Object, e As System.EventArgs) Handles cmdSend.Click
        ' Make sure that the users send no empty messages
        If rtfSend.Text.Length > 0 Then
            Dim msg = New Matrix.Xmpp.Client.Message() With { _
                    .Type = MessageType.groupchat, _
                    .[To] = _roomJid, _
                    .Body = rtfSend.Text _
                    }

            _xmppClient.Send(msg)

            rtfSend.Clear()
        End If
    End Sub

    ''' <summary>
    ''' Changing the subject in a chatroom in MUC rooms this could return an error when you are a normal user and not allowed
    ''' to change the subject.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdChangeSubject_Click(sender As Object, e As System.EventArgs) Handles cmdChangeSubject.Click
        Dim msg = New Matrix.Xmpp.Client.Message() With { _
                .Type = MessageType.groupchat, _
                .[To] = _roomJid, _
                .Subject = txtSubject.Text _
                }

        _xmppClient.Send(msg)
    End Sub
End Class

File: MatriX/samples/csharp/MiniClient/FrmGroupChat.cs
using System.Drawing;
using System.Windows.Forms;

using Matrix;
using Matrix.Xmpp;
using Matrix.Xmpp.Client;
using Matrix.Xmpp.Muc.User;
using EventArgs = System.EventArgs;

namespace MiniClient
{
    public partial class FrmGroupChat : Form
    {
        private MucManager mm;
        #region << Constructors >>        
        public FrmGroupChat(XmppClient xmppClient, Jid roomJid, string nickname)
        {
            InitializeComponent();   
            
            _roomJid = roomJid;
            _xmppClient = xmppClient;
            _nickname = nickname;

            mm = new MucManager(xmppClient);

            // Setup new Message Callback using the MessageFilter
            _xmppClient.MessageFilter.Add(roomJid, new BareJidComparer(), MessageCallback);
            
            // Setup new Presence Callback using the PresenceFilter
            _xmppClient.PresenceFilter.Add(roomJid, new BareJidComparer(), PresenceCallback);

        }
        #endregion

        private readonly Jid                     _roomJid;
        private readonly XmppClient              _xmppClient;
        private readonly string                  _nickname;

        private void FrmGroupChat_Load(object sender, EventArgs e)
        {
            if (_roomJid != null)
            {
                mm.EnterRoom(_roomJid, _nickname);
            }
        }

        private void FrmGroupChat_FormClosed(object sender, FormClosedEventArgs e)
        {
            if (_roomJid != null)
            {
                mm.ExitRoom(_roomJid, _nickname);
                
                // Remove the Message Callback in the MessageFilter
                _xmppClient.MessageFilter.Remove(_roomJid);

                // Remove the Presence Callback in the MessageFilter
                _xmppClient.PresenceFilter.Remove(_roomJid);
            }

        }
       
        private void MessageCallback(object sender, MessageEventArgs e)
        {
            if (e.Message.Type == MessageType.groupchat)
                IncomingMessage(e.Message);
        }

      
        private void PresenceCallback(object sender, PresenceEventArgs e)
        {

            var mucX = e.Presence.MucUser;
            
            // check for status code 201, this means the room is not ready to use yet
            // we request an instant room and accept the and accept the default configuration by the server
            if (mucX.HasStatus(201)) // 201 =  room is awaiting configuration.
                mm.RequestInstantRoom(_roomJid);
           

            var lvi = FindListViewItem(e.Presence.From);
            if (lvi != null)
            {
                if (e.Presence.Type == PresenceType.unavailable)
                {
                    lvi.Remove();
                }
                else
                {
                    int imageIdx = Util.GetRosterImageIndex(e.Presence);
                    lvi.ImageIndex = imageIdx;
                    lvi.SubItems[1].Text = (e.Presence.Status ?? "");
                    
                    var u = e.Presence.MucUser;
                    if (u != null)
                    {
                        lvi.SubItems[2].Text = u.Item.Affiliation.ToString();
                        lvi.SubItems[3].Text = u.Item.Role.ToString();
                    }
                }
            }
            else
            {
                int imageIdx = Util.GetRosterImageIndex(e.Presence);
                
                var lv = new ListViewItem(e.Presence.From.Resource) {Tag = e.Presence.From.ToString()};

                lv.SubItems.Add(e.Presence.Status ?? "");
                
                var u = e.Presence.MucUser;
                if (u != null)
                {
                    lv.SubItems.Add(u.Item.Affiliation.ToString());
                    lv.SubItems.Add(u.Item.Role.ToString());
                }
                lv.ImageIndex = imageIdx;
                lvwRoster.Items.Add(lv);
            }
        }

        private ListViewItem FindListViewItem(Jid jid)
        {
            foreach (ListViewItem lvi in lvwRoster.Items)
            {
                if (jid.ToString().ToLower() == lvi.Tag.ToString().ToLower())
                    return lvi;
            }
            return null;
        }

        private void IncomingMessage(Matrix.Xmpp.Client.Message msg)
        {
            if (msg.Type == MessageType.error)
            {
                //Handle errors here
                // we dont handle them in this example
                return;
            }

            if (msg.Subject != null)
            {
                txtSubject.Text = msg.Subject;

                rtfChat.SelectionColor = Color.DarkGreen;
                // The Nickname of the sender is in GroupChat in the Resource of the Jid
                rtfChat.AppendText(msg.From.Resource + " changed subject: ");
                rtfChat.SelectionColor = Color.Black;                
                rtfChat.AppendText(msg.Subject);
                rtfChat.AppendText("\r\n");
            }
            else
            {
                if (msg.Body == null)
                    return;

                rtfChat.SelectionColor = Color.Red;
                // The Nickname of the sender is in GroupChat in the Resource of the Jid
                rtfChat.AppendText(msg.From.Resource + " said: ");
                rtfChat.SelectionColor = Color.Black;
                rtfChat.AppendText(msg.Body);
                rtfChat.AppendText("\r\n");
            }
        }

        private void cmdSend_Click(object sender, EventArgs e)
        {
            // Make sure that the users send no empty messages
            if (rtfSend.Text.Length > 0)
            {
                var msg = new Matrix.Xmpp.Client.Message
                              {
                                  Type = MessageType.groupchat,
                                  To = _roomJid,
                                  Body = rtfSend.Text
                              };

                _xmppClient.Send(msg);

                rtfSend.Clear();
            }
        }

        /// <summary>
        /// Changing the subject in a chatroom in MUC rooms this could return an error when you are a normal user and not allowed
        /// to change the subject.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdChangeSubject_Click(object sender, EventArgs e)
        {
            var msg = new Matrix.Xmpp.Client.Message
                          {
                              Type = MessageType.groupchat,
                              To = _roomJid,
                              Subject = txtSubject.Text
                          };

            _xmppClient.Send(msg);
        }
    }
}